source/darwin/nw_socket.c

```diff
@@ -1535,8 +2068,15 @@ static int s_socket_listen_fn(struct aws_socket *socket, int backlog_size) {
         return aws_raise_error(AWS_IO_SOCKET_ILLEGAL_OPERATION_FOR_STATE);
     }
 
-    socket->io_handle.data.handle = nw_listener_create(nw_socket->socket_options_to_params);
+    if (nw_socket->nw_parameters == NULL) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p: socket nw_parameters needs to be set before creating a listener from socket.",
+            (void *)socket);
+        return aws_raise_error(AWS_IO_SOCKET_INVALID_OPTIONS);
```

Comment:
We still hold the lock here on this early out.  I'd add a done label at the bottom that returns a result variable after unlocking and switch all the early outs to gotos